From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 10 Jul 2023 22:05:08 +0200
Subject: [PATCH] Modifiable Bukkit enums - Remove hard-coded properties -
 Material - Key

Modifies the `Material` property `getKey()` so that it is no longer computed in `Material`, but based on a function defined and implemented in `MaterialNameAndKeyConversion`.

The property values are cached after being computed, for performant access.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 0a8e4c247d802feedcdeee8b8dcbed5d58f807fd..ea8c593826811031046d8858cf770e467b8ccd1e 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4393,7 +4393,6 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     private final short durabilityHardcoded;
     public final Class<?> data;
     private final boolean legacy;
-    private final NamespacedKey key;
 
     private Material(final int id) {
         this(id, 64);
@@ -4421,7 +4420,6 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         this.maxStackHardcoded = stack;
         this.data = data;
         this.legacy = this.name().startsWith(LEGACY_PREFIX);
-        this.key = NamespacedKey.minecraft(this.name().toLowerCase(Locale.ROOT));
         // try to cache the constructor for this material
         try {
             if (MaterialData.class.isAssignableFrom(data)) {
@@ -4520,8 +4518,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     @NotNull
     @Override
     public NamespacedKey getKey() {
-        Preconditions.checkArgument(!legacy, "Cannot get key of Legacy Material");
-        return key;
+        return org.fiddlemc.fiddle.material.MaterialNameAndKeyConversion.getKey(this); // Fiddle - modifiable Bukkit enums - Material - remove hard-coded properties - key
     }
 
     /**
diff --git a/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java b/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
index db0b953c6aacd29817dcb0bd5069346ac7116b80..b600f06d62ea4b045590d72626d8029be82e89f2 100644
--- a/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
+++ b/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
@@ -2,7 +2,12 @@
 
 package org.fiddlemc.fiddle.material;
 
+import com.google.common.base.Preconditions;
 import org.bukkit.Material;
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Locale;
 
 /**
  * This class provides a number of implementations for the conversion of names and keys
@@ -17,4 +22,20 @@ public final class MaterialNameAndKeyConversion {
         throw new RuntimeException();
     }
 
+    // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - key
+    private static final CachedMaterialProperty<NamespacedKey> keyCache = new CachedMaterialProperty<>(
+        NamespacedKey[]::new,
+        // Based on the original org.bukkit.Material constructor
+        material -> NamespacedKey.minecraft(material.name().toLowerCase(Locale.ROOT))
+    );
+
+    /**
+     * @throws IllegalArgumentException If the given {@code material} is {@linkplain Material#isLegacy() legacy}.
+     */
+    public static @NotNull NamespacedKey getKey(Material material) {
+        Preconditions.checkArgument(!material.isLegacy(), "Cannot get key of Legacy Material");
+        return keyCache.get(material);
+    }
+    // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - key
+
 }
diff --git a/src/test/java/org/bukkit/MaterialTest.java b/src/test/java/org/bukkit/MaterialTest.java
index 6b0eaaadf9772a4850cb70cea2bfe72a81a47a04..ae3766e1804ca495ee1f5e1ee77cd1b4fee916b8 100644
--- a/src/test/java/org/bukkit/MaterialTest.java
+++ b/src/test/java/org/bukkit/MaterialTest.java
@@ -52,6 +52,18 @@ public class MaterialTest {
         }
     }
 
+    // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - key
+    @Test
+    public void matchMaterialByKeyPath() {
+        for (Material material : Material.values()) {
+            if (material.isLegacy()) {
+                continue;
+            }
+            assertThat(Material.matchMaterial(material.getKey().getKey()), is(material));
+        }
+    }
+    // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - key
+
     @Test
     public void matchMaterialByWrongNamespace() {
         for (Material material : Material.values()) {
